x-pack/plugins/cases/server/connectors/cases/README.md

```diff
@@ -0,0 +1,579 @@
+top --> bottom
+```
+
+## Grouping
+
+The case action accepts an array of alerts provided by the connector adapter. Duplicate alerts will not be attached to the same case. The case action groups the alerts by the grouping field configured by the user. For example, if the grouping field is `host.name` the case action will group the alerts by the values of the `host.name` field. Users may define more than one grouping field. In this case, the grouping will be done by multiple fields. The grouping is performed in memory by the case action as the number of alerts is expected to be low on average and they are already loaded in memory by the alerting framework.
+
+```mermaid
+flowchart LR
+
+    alert1 --> groupingBy
+    alert2 --> groupingBy
+    alert3 --> groupingBy
+
+    subgraph top ["IP: 0.0.0.1"]
+        alert11[Alert 1]
+        alert22[Alert 2]
+    end
+
+    subgraph bottom ["IP: 0.0.0.2"]
+        alert23[Alert 3]
+    end
+
+    groupingBy --> top
+    groupingBy --> bottom
+
+    alert1[Alert 1]
+    alert2[Alert 2]
+    alert3[Alert 3]
+    groupingBy[Grouping by IP]
+```
+
+```mermaid
+flowchart LR
+
+    alert1 --> groupingBy
+    alert2 --> groupingBy
+    alert3 --> groupingBy
+
+    subgraph top ["IP: 0.0.0.1 AND host.name: A"]
+        alert11[Alert 1]
+        alert22[Alert 2]
+    end
+
+    subgraph bottom ["IP: 0.0.0.2 AND host.name: B"]
+        alert23[Alert 3]
+    end
+
+    groupingBy --> top
+    groupingBy --> bottom
+
+    alert1[Alert 1]
+    alert2[Alert 2]
+    alert3[Alert 3]
+    groupingBy[Grouping by IP and host name]
+```
+
+## Case creation
+
+For each rule and each group produced by the grouping step, a case will be created and the alerts of that group will be attached to the new case. In future executions of the rule, new alerts that belong to the same group will be attached to the same case. If an alert cannot be grouped, because the grouping field does not exist in its data, it will be attached to a case that represents the `unknown` value.
+
+To support this, the case action constructs a deterministic deduplication ID which will be set as the case ID. The ID can be constructed on each execution of the case action without the need to persist it and can correctly map alerts of the same group to the case that represents that group. A deduplication ID has two main advantages:
+
+1. The case action can determine, without persisting the case ID, if a case exists for a specific rule and a specific group.
+2. If two Kibana nodes or two executions of the same rule try to create the case with the same ID only one node or execution will succeed. See the Race Conditions section for more details.
+
+The deduplication ID will be constructed as 
+
+`sha256(<rule_id>:<space_id>:<owner>:<grouping_definition>:<counter>)`
+
+where
+
+`<rule_id>`: The ID of the rule. It can be optional to support two different rules attaching alerts to the same case. The rule id must be included if the grouping definition is not defined.
```

Comment:
> It can be optional to support two different rules attaching alerts to the same case.

This isn't supported(yet?) so maybe we should remove the line from the documentation.

> The rule id must be included if the grouping definition is not defined.

Isn't the rule id **always** included anyway? this part might also be unnecessary.

How about something like:
```suggestion
`<rule_id>`: The ID of the rule. Including it in the deduplication ID ensures all cases created are at least rule-specific.
```